source/common/http/conn_manager_impl.cc

```diff
@@ -782,6 +782,19 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(ActiveStreamDecoderFilte
     ENVOY_STREAM_LOG(trace, "decode headers called: filter={} status={}", *this,
                      static_cast<const void*>((*entry).get()), static_cast<uint64_t>(status));
 
+    if (!request_metadata_map_vector_.empty()) {
+      for (const auto& metadata_map : request_metadata_map_vector_) {
+        decodeMetadata(nullptr, *metadata_map);
+      }
+      // If metadata is added to a header only request, we should add an empty data frame to
+      // indicate the end of the stream.
+      if ((*entry)->end_stream_) {
+        Buffer::OwnedImpl empty_data("");
+        addDecodedData(*((*entry).get()), empty_data, true);
```

Comment:
Apologize for the long response!

Added a trace log.

Great catch!! It wasn't working with decoding_headers_only_! To solve this problem, I tried to pass the empty data frame (added by addDecodedData) through filters when decoding_headers_only_ is true, but failed. The main problem is hard to distinguish this empty data frame with the data frames that are received from downstream (which we will ignore because of decoding_headers_only_). We could add more conditions to allow empty data frame to go through the filters here: https://github.com/envoyproxy/envoy/blob/208a7fc72dd5f63fe1722111a248de4ce9224eab/source/common/http/conn_manager_impl.cc#L822. But then I can't think of a way to set end_stream true in the empty data frame. I think complete() has to be true in order to set the end_stream in the empty data frame (https://github.com/envoyproxy/envoy/blob/208a7fc72dd5f63fe1722111a248de4ce9224eab/source/common/http/conn_manager_impl.cc#L1529). However, if I set complete() to be true, the newly received data frame from downstream will cause assert failure (complete() is true means we are done with downstream, shouldn't receive more data).

A very hacky solution I used is to add an empty data in router filter when end_stream in headers is true and we still have metadata to send. The downside is the data frame won't go through filters (Since filters decides to process headers only, it might be OK). The upside is it's simple.

But now the code is messy: if metadata is added in headers only request, the added empty data frame will go through filters. If metadata is added with decoding_headers_only_ true, the added empty data will not go through filters. I could merge those two cases to one: only add new empty data in router filters. What do you think?

About test: GrpcHttp1ReverseBridge returns ContinueAndEndStream in encodeHeaders(). So I used decode-headers-only to test, which returns ContinueAndEndStream in decodeHeaders(). Did I miss anything?

Please let me know any suggestions/thoughts. Thanks!!

Reply:
Metadata is sent as h2 extension frames. Some implementations may ignore or mishandle this frame type. It could potentially cause problems if we allow metadata to carry end_stream. But if we think it is not a concern, I can change to allow end_stream. If we do so, we may differ from GFE implementation.